### PR TITLE
setup autolinking for packages in bare-expo

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Simplify the pod and gradle configurations for the native runner. This will help us as we work towards managed EAS support, because the project will be closer to the template.

### CODE REVIEW

File: packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js [hunk 1]
```diff
@@ -16,6 +16,12 @@ function symlinkNecessaryPackages(projectPath) {
 
   symlinkNecessaryPackage(projectPath, 'expo');
   symlinkNecessaryPackage(projectPath, 'react-native');
+  // For community CLI autolinking in bare-expo
+  symlinkNecessaryPackage(projectPath, 'expo-image');
+  symlinkNecessaryPackage(projectPath, 'expo-random');
+  symlinkNecessaryPackage(projectPath, 'expo-dev-launcher');
+  symlinkNecessaryPackage(projectPath, 'expo-dev-menu');
+  symlinkNecessaryPackage(projectPath, 'expo-dev-menu-interface');
```
Comment (lukmccall) [inline packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:24]: I'm not sure if we need to create symlinks for those packages. RN CLI link should work without that. Or maybe there is another reason why you add that?

Comment (EvanBacon) [reply packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:24]: afaict without these symlinks, the packages don't get autolinked to bare-expo. Can test by removing, running yarn, and running pod install in `apps/bare-expo/ios`. The packages will then be uninstalled.

Comment (ide) [reply packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:26]: (a) keep code specific to bare-expo out of expo-yarn-workspaces. can put it in bare-expo instead.
(b) what does the autolinking code in the RN CLI do? does it only look in the project root's node_modules?

Comment (lukmccall) [reply packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:26]: Note that the `expo-dev-menu-interface` isn't listed in the podfile now, but it still is linked into the project (without the symlink). 

Moreover, the RN config lists the package correctly on both platforms: 
```
    "expo-dev-menu-interface": {
      "root": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface",
      "name": "expo-dev-menu-interface",
      "platforms": {
        "ios": {
          "sourceDir": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface/ios/EXDevMenuInterface.xcodeproj",
          "folder": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface",
          "pbxprojPath": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface/ios/EXDevMenuInterface.xcodeproj/project.pbxproj/project.pbxproj",
          "podfile": null,
          "podspecPath": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface/EXDevMenuInterface.podspec",
          "projectPath": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface/ios/EXDevMenuInterface.xcodeproj/project.pbxproj",
          "projectName": "project.pbxproj",
          "libraryFolder": "Libraries",
          "sharedLibraries": [],
          "plist": [],
          "scriptPhases": []
        },
        "android": {
          "sourceDir": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface/android",
          "folder": "/Users/lukasz/work/expo/packages/expo-dev-menu-interface",
          "packageImportPath": "import expo.interfaces.devmenu.DevMenuInterfacePackage;",
          "packageInstance": "new DevMenuInterfacePackage()"
        }
      },
      "assets": [],
      "hooks": {},
      "params": []
    },
```

Comment (lukmccall) [reply packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:26]: but maybe after you merged https://github.com/expo/expo/pull/11402, it stopped working. If that's the case, I'm ok with these changes ;)

File: apps/bare-expo/ios/Podfile [hunk 2]
```diff
@@ -28,11 +28,6 @@ abstract_target 'BareExpoMain' do
 
   use_react_native!(path: config['reactNativePath'])
 
-  pod 'EXDevMenu', path: '../../../packages/expo-dev-menu', configurations: :debug
```
Comment (lukmccall) [inline apps/bare-expo/ios/Podfile:31]: This line was added to remove the `expo-dev-menu` from release builds and I think we should leave that.

Comment (EvanBacon) [reply apps/bare-expo/ios/Podfile:31]: hmm we should have some heuristic for this functionality that can be statically generated for managed workflow support. Any ideas?

**Review by ide — CHANGES_REQUESTED** [review]
Requesting changes to move the symlinking out of expo-yarn-workspaces, which is designed to be a little more general than for this repo's structure, and into bare-expo.

File: packages/expo-dev-launcher/react-native.config.js [hunk 3]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (ide) [inline packages/expo-dev-launcher/react-native.config.js:1]: Is this supposed to be JSON?

File: packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js [hunk 4]
```diff
@@ -15,13 +15,30 @@ function symlinkNecessaryPackages(projectPath) {
...
   symlinkNecessaryPackage(projectPath, 'react-native');
-  symlinkNecessaryPackage(projectPath, 'react-native-unimodules');
-  symlinkNecessaryPackage(projectPath, '@react-native-community/cli-platform-ios');
-  symlinkNecessaryPackage(projectPath, '@react-native-community/cli-platform-android');
+ 
+  const packageJson = getProjectPackageJson(projectPath);
+
+  const workspaces = packageJson.workspaces || {};
+  const symlinks = workspaces.symlinks || [];
+  debug(`Project defined symlinks `, symlinks);
```
Comment (ide) [inline packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:26]:
```suggestion
  debug(`Project defined symlinks`, symlinks);
```

File: packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js [hunk 5]
```diff
@@ -15,13 +15,30 @@ function symlinkNecessaryPackages(projectPath) {
...
+    }
+  }
+}
+
+function getProjectPackageJson(projectPath) { 
+  try {
+    const contents = fs.readFileSync(path.join(projectPath, 'package.json'));
+    return JSON.parse(contents);
+  } catch {
+    return {}
```
Comment (ide) [inline packages/expo-yarn-workspaces/bin/symlink-necessary-packages.js:40]:
```suggestion
    return {};
```

File: packages/expo-yarn-workspaces/README.md [hunk 6]
```diff
@@ -39,3 +39,11 @@ module.exports = createMetroConfiguration(__dirname);
 The `expo-yarn-workspaces` package defines a Metro configuration object that makes Metro work with Yarn workspaces in the Expo repo. It configures Metro to include packages from the workspace root, resolves symlinked packages, excludes modules from Haste's module system, and exclude modules in the native Android and Xcode projects. You can further customize this configuration object before exporting it, if needed.
 
 **Aside:** when starting the project, run `expo start --clear` so Metro uses the latest configuration instead of working with cached values.
+
+### Configuration
+
+You can configure workspaces using the `workspaces` object in the `package.json`.
```
Comment (ide) [inline packages/expo-yarn-workspaces/README.md:45]:
```suggestion
You can configure workspaces using the `workspace` field in each workspace package's package.json file.
```

File: packages/expo-yarn-workspaces/README.md [hunk 7]
```diff
@@ -39,3 +39,11 @@ module.exports = createMetroConfiguration(__dirname);
...
+
+#### symlinks
+
+Sometimes a node module must be located in the project's `node_modules` folder for things to work properly. Using the `workspaces.symlinks` string array you can define a list of packages that should be linked on install.
```
Comment (ide) [inline packages/expo-yarn-workspaces/README.md:49]:
```suggestion
Sometimes an npm package must be located in the project's `node_modules` folder for things to work properly. Using the `workspaces.symlinks` string array you can define a list of packages to symlink under the project's `node_modules` folder after installing the workspaces' dependencies.

**Review by ide — CHANGES_REQUESTED** [review]

File: packages/expo-dev-launcher/react-native.config.js (hunk 3 above)
Comment (EvanBacon) [reply packages/expo-dev-launcher/react-native.config.js:1]: copied over from what brent implemented in expo-random

Comment (ide) [reply packages/expo-dev-launcher/react-native.config.js:1]: That file might not be right: https://github.com/react-native-community/cli/blob/master/docs/platforms.md#how-does-it-work

**Review by ide — CHANGES_REQUESTED** [review]
From what I can see, react-native.config.js is supposed to be a JS module: https://github.com/react-native-community/cli/blob/master/docs/platforms.md#how-does-it-work

File: apps/bare-expo/ios/Podfile (hunk 2 above)
Comment (lukmccall) [reply apps/bare-expo/ios/Podfile]: We'll remove it after the next publish of the RN CLI -  https://github.com/react-native-community/cli/pull/1298

File: package.json [hunk 8]
```diff
@@ -11,7 +11,7 @@
     "lint": "eslint .",
     "tsc": "echo 'You are trying to run \"tsc\" in the workspace root. Run it from an individual package instead.' && exit 1"
   },
-  "workspaces": {
+  "workspace": {
```
Comment (EvanBacon) [inline package.json:14]:
```suggestion
  "workspaces": {
```

File: packages/expo-yarn-workspaces/README.md [hunk 9]
```diff
@@ -39,3 +39,11 @@ module.exports = createMetroConfiguration(__dirname);
 The `expo-yarn-workspaces` package defines a Metro configuration object that makes Metro work with Yarn workspaces in the Expo repo. It configures Metro to include packages from the workspace root, resolves symlinked packages, excludes modules from Haste's module system, and exclude modules in the native Android and Xcode projects. You can further customize this configuration object before exporting it, if needed.
 
 **Aside:** when starting the project, run `expo start --clear` so Metro uses the latest configuration instead of working with cached values.
+
+### Configuration
+
+You can configure workspaces using the `workspace` field in each workspace package's package.json file.
```
Comment (ide) [inline packages/expo-yarn-workspaces/README.md:45]:
```suggestion
You can configure workspaces using the `expo-yarn-workspaces` field in each workspace package's package.json file.
```

File: packages/expo-yarn-workspaces/README.md [hunk 10]
```diff
@@ -39,3 +39,11 @@ module.exports = createMetroConfiguration(__dirname);
...
+
+#### symlinks
+
+Sometimes an npm package must be located in the project's `node_modules` folder for things to work properly. Using the `workspace.symlinks` string array you can define a list of packages to symlink under the project's `node_modules` folder after installing the workspaces' dependencies.
```
Comment (ide) [inline packages/expo-yarn-workspaces/README.md:49]:
```suggestion
Sometimes an npm package must be located in the project's `node_modules` folder for things to work properly. Using the `expo-yarn-workspace.symlinks` string array you can define a list of packages to symlink under the project's `node_modules` folder after installing the workspaces' dependencies.
```

File: packages/expo-dev-launcher/react-native.config.js (hunk 3 above)
Comment (ide) [reply packages/expo-dev-launcher/react-native.config.js:1]: Seems like this still needs to be fixed

**Review by ide — CHANGES_REQUESTED** [review]

**Pull request closed without merging**